src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -85,23 +102,113 @@ protected override Payload ConstructHeadersPayload(BlockLocator locator, uint256
+                this.consensusManager,
+                this.peerBanning,
+                this.loggerFactory,
+                this.network,
+                this.chainState);
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer supports Proven Headers and PH has been activated.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if is peer is PH enabled; otherwise, <c>false</c>.
+        /// </returns>
+        private bool CanPeerProcessProvenHeaders(INetworkPeer peer)
+        {
+            return peer.Version >= NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION;
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer is Whitelisted.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if the specified peer is Whitelisted; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerWhitelisted(INetworkPeer peer)
+        {
+            return peer.Behavior<IConnectionManagerBehavior>()?.Whitelisted == true;
+        }
+
+        private bool AreProvenHeadersActivated()
+        {
+            if (this.network.Consensus.Options is PosConsensusOptions options)
+            {
+                long currentHeight = this.chainState.ConsensusTip.Height;
+                return currentHeight >= options.ProvenHeadersActivationHeight;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Builds the <see cref="GetHeadersPayload"/>.
+        /// </summary>
+        /// <returns>The <see cref="GetHeadersPayload"/> instance.
+        /// If the peer can serve PH, <see cref="GetProvenHeadersPayload" /> is returned, otherwise if it's a legacy peer but it's whitelisted,
+        /// <see cref="GetHeadersPayload"/> is returned.
+        /// If the attached peer is a legacy peer and it's not whitelisted, returns null.
+        /// </returns>
+        protected override GetHeadersPayload BuildGetHeadersPayload()
+        {
+            INetworkPeer peer = this.AttachedPeer;
+
+            if (this.AreProvenHeadersActivated())
+            {
+                if (this.CanPeerProcessProvenHeaders(peer))
+                {
+                    return new GetProvenHeadersPayload()
+                    {
+                        BlockLocator = (this.ExpectedPeerTip ?? this.consensusManager.Tip).GetLocator(),
+                        HashStop = null
+                    };
+                }
+                // If the peer doesn't supports PH but it's whitelisted, issue a standard GetHeadersPayload
+                else if (this.IsPeerWhitelisted(peer))
+                    return base.BuildGetHeadersPayload();
+                // If the peer doesn't support PH and isn't whitelisted, return null (stop synch attempt with legacy StratisX nodes).
+                else
```

Comment:
Its a bit confusing about what this `if` this `else` belongs to so maybe add braces?

Reply:
oh if it was for me I'd add always braces even on single-line blocks!
Braces added, let's merge